[libc++] Fix checks for terminal and flushes in std::print()

dimztimz opened the pull request:
The check whether a stream is associated with a terminal or not and the flushing of the stream in `std::print()` is needed only on Windows. Additionally, the correct flush should be used. When `std::print` is called with a C stream, `std::fflush()` should be used. When it is called with C++ `ostream`, `ostream::flush()` should be called.

Because POSIX does not have a separate Unicode API for terminal output, checking for terminal (`isatty`) and flushing is not needed at all. Moreover, `isatty` has noticeable performance cost.

See also https://wg21.link/LWG4044.

Fixes #70142

philnik777 commented:
@vitaut would you mind taking a look too?

vitaut commented:
The change looks correct to me. I haven't looked at the tests though.

dimztimz commented:
I updated this PR so it solves the bug for the ostream overloads too. The LWG issue is in state "tentatively ready", IMO good enough to accept this.

mordante commented:
I had a look and I think this is not a bug in libc++; it's a bug in the standard. 
https://cplusplus.github.io/LWG/issue4044
This LWG issue will likely be voted in during November 23 plenary.
This is well in time for the LLVM-20 release. So I'll add the LLVM-20 milestone for now.

dimztimz commented:
> I had a look and I think this is not a bug in libc++; it's a bug in the standard.
> https://cplusplus.github.io/LWG/issue4044

IMO it is a bug in the library even with the current wording, as I explained in the linked issue. The new wording is definitely better, but you do not need to wait for it to get accepted. GCC's behavior is as if the issue is accepted. Anyway, I rebased this again.

mordante commented:
> @mordante What's the status of this patch? Should we go ahead with it or close it?

We should go ahead with it, but not now. I actually wrote the information about its status yesterday https://github.com/llvm/llvm-project/issues/70142#issuecomment-2750480205

dimztimz committed:
[libc++] Fix checks for terminal and flushes in std::print()

The check whether a stream is associated with a terminal or not and the
flushing of the stream in std::print() is needed only on Windows.
Additionally, the correct flush should be used. When std::print is
called with a C stream, std::fflush() should be used. When it is called
with C++ ostream, ostream::flush() should be called.

Because POSIX does not have a separate Unicode API for terminal output,
checking for terminal (isatty) and flushing is not needed at all.
Moreover, isatty has noticeable performance cost.

See also https://wg21.link/LWG4044.

Fixes #70142

dimztimz commented on libcxx/include/print:
Yes, it is dead it is never called. Looking at _`__is_terminal`, I think that it can be refactored and simplified, I see some compile-time config `_LIBCPP_HAS_TERMINAL` that I don't think it is needed. I will close this, let's keep this cleaning and refactoring for another issue. IMO, this PR should be only about the bug.

dimztimz commented on libcxx/include/print:
```diff
@@ -316,9 +301,15 @@ __vprint_unicode([[maybe_unused]] FILE* __stream,
-  __print::__vprint_unicode_windows(__stream, __fmt, __args, __write_nl, __print::__is_terminal(__stream));
+  if (__print::__is_terminal(__stream)) {
+    // TODO PRINT Should flush errors throw too?
+    std::fflush(__stream);
```
Here is the flush.

dimztimz commented on libcxx/include/print:
```diff
@@ -236,26 +236,11 @@ __vprint_nonunicode(FILE* __stream, string_view __fmt, format_args __args, bool
-    return __print::__vprint_nonunicode(__stream, __fmt, __args, __write_nl);
-
-  // TODO PRINT Should flush errors throw too?
-  std::fflush(__stream);
```
The answer why the flush is needed at all is in the paper https://wg21.link/p2539r4.

dimztimz replied:
And the LWG issue https://wg21.link/lwg4044.

dimztimz committed:
fixup! [libc++] Fix checks for terminal and flushes in std::print()

simplify and add ifdef guards to __is_terminal

dimztimz committed:
fixup! [libc++] Fix checks for terminal and flushes in std::print()

drop _LIBCPP_HAS_TERMINAL compile-time config

philnik777 reviewed (commented):
Mostly LGTM. Just some minor comments.

dimztimz committed:
fixup! [libc++] Fix checks for terminal and flushes in std::print()

drop AVAILABILITY define, do not modify .gn file

philnik777 reviewed (approved):
Thanks! Do you have commit access, or should I merge the patch?

dimztimz commented:
No direct push access, but I think I should have access to merge button via Github UI. It's still not there. Should we wait for the CI maybe?

philnik777 commented:
> No direct push access, but I think I should have access to merge button via Github UI. It's still not there. Should we wait for the CI maybe?

I think that means you don't have access. We should definitely  wait for the CI though.
